File: go/ch03/score_module.go
// go/ch03/score_module.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

type ProblemInput struct {
	Days  int
	Decay []int
	Gain  [][]int
}

type ScheduleState struct {
	Types []int
}

type ScoreResult struct {
	Daily              []int
	FinalSatisfaction  int
	ContestScore       int
}

func readProblemWithSchedule(r io.Reader) (ProblemInput, ScheduleState, error) {
	sc := bufio.NewScanner(r)
	var lines []string
	for sc.Scan() {
		lines = append(lines, sc.Text())
	}
	if err := sc.Err(); err != nil {
		return ProblemInput{}, ScheduleState{}, err
	}

	idx := 0
	d, _ := strconv.Atoi(lines[idx])
	idx++
	decay := parseInts(lines[idx])
	idx++
	gain := make([][]int, d)
	for i := 0; i < d; i++ {
		gain[i] = parseInts(lines[idx])
		idx++
	}
	types := make([]int, d)
	for i := 0; i < d; i++ {
		types[i], _ = strconv.Atoi(lines[idx])
		idx++
	}
	return ProblemInput{d, decay, gain}, ScheduleState{types}, nil
}

func parseInts(line string) []int {
	parts := strings.Fields(line)
	out := make([]int, len(parts))
	for i, p := range parts {
		out[i], _ = strconv.Atoi(p)
	}
	return out
}

func applyMove(state ScheduleState, day, contestType int) ScheduleState {
	nxt := append([]int{}, state.Types...)
	nxt[day] = contestType
	return ScheduleState{nxt}
}

func scoreSchedule(problem ProblemInput, state ScheduleState) ScoreResult {
	last := make([]int, 26)
	satisfaction := 0
	daily := make([]int, 0, problem.Days)

	for day := 1; day <= problem.Days; day++ {
		t := state.Types[day-1] - 1
		satisfaction += problem.Gain[day-1][t]
		last[t] = day
		decay := 0
		for i := 0; i < 26; i++ {
			decay += problem.Decay[i] * (day - last[i])
		}
		satisfaction -= decay
		daily = append(daily, satisfaction)
	}

	final := 0
	if len(daily) > 0 {
		final = daily[len(daily)-1]
	}
	contestScore := 1000000 + final
	if contestScore < 0 {
		contestScore = 0
	}
	return ScoreResult{daily, final, contestScore}
}

func main() {
	problem, state, err := readProblemWithSchedule(os.Stdin)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	result := scoreSchedule(problem, state)
	for _, v := range result.Daily {
		fmt.Println(v)
	}
	fmt.Fprintln(os.Stderr, result.ContestScore)
}

File: python/ch03/score_module.py
# python/ch03/score_module.py
"""Intro HC 形式のスコア計算モジュール（第3章）."""
from __future__ import annotations

import sys
from dataclasses import dataclass
from typing import List, TextIO, Tuple


# --- 3.3 State ---


@dataclass
class ProblemInput:
    days: int
    decay: List[int]
    gain: List[List[int]]


@dataclass
class ScheduleState:
    types: List[int]  # 1-indexed contest types

    def copy(self) -> ScheduleState:
        return ScheduleState(list(self.types))


@dataclass
class ScoreResult:
    daily: List[int]
    final_satisfaction: int
    contest_score: int


# --- 3.1 read ---


def read_problem(stream: TextIO) -> ProblemInput:
    lines = stream.read().splitlines()
    idx = 0
    days = int(lines[idx])
    idx += 1
    decay = list(map(int, lines[idx].split()))
    idx += 1
    gain = []
    for _ in range(days):
        gain.append(list(map(int, lines[idx].split())))
        idx += 1
    return ProblemInput(days=days, decay=decay, gain=gain)


def read_problem_with_schedule(stream: TextIO) -> Tuple[ProblemInput, ScheduleState]:
    lines = stream.read().splitlines()
    idx = 0
    days = int(lines[idx])
    idx += 1
    decay = list(map(int, lines[idx].split()))
    idx += 1
    gain = []
    for _ in range(days):
        gain.append(list(map(int, lines[idx].split())))
        idx += 1
    types = [int(lines[idx + i]) for i in range(days)]
    return ProblemInput(days, decay, gain), ScheduleState(types)


# --- 3.4 score ---


def apply_move(state: ScheduleState, day: int, contest_type: int) -> ScheduleState:
    """day: 0-indexed, contest_type: 1-indexed."""
    nxt = state.copy()
    nxt.types[day] = contest_type
    return nxt


def score_schedule(problem: ProblemInput, state: ScheduleState) -> ScoreResult:
    last = [0] * 26
    satisfaction = 0
    daily: List[int] = []

    for day in range(1, problem.days + 1):
        t = state.types[day - 1] - 1
        satisfaction += problem.gain[day - 1][t]
        last[t] = day
        decay = sum(problem.decay[i] * (day - last[i]) for i in range(26))
        satisfaction -= decay
        daily.append(satisfaction)

    final = daily[-1] if daily else 0
    return ScoreResult(
        daily=daily,
        final_satisfaction=final,
        contest_score=max(10**6 + final, 0),
    )


def main() -> None:
    problem, state = read_problem_with_schedule(sys.stdin)
    result = score_schedule(problem, state)
    for v in result.daily:
        print(v)
    print(result.contest_score, file=sys.stderr)


if __name__ == "__main__":
    main()
